keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1267 @@
+	// status reason. The condition is satisfied when the actual value of the
+	// status.reason field does not equal any of the listed values.
+	// +optional
+	PodStatusReasonNotIn *[]string
+}
+
+// BackoffPolicy describes the handling of failed pods. The Rules field specifies
+// the list of rules. Each rule specifies an action to be taken on a pod
+// failure and a condition to be checked if the rule applies.
+type BackoffPolicy struct {
+	Rules []BackoffPolicyRule
+}
+
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+  ...
+	// Specifies the policy of handling failed pods. In particular, it allows to
+	// specify the set of actions and conditions which need to be
+	// satisfied to take the associated action.
+	// If empty, the default behaviour applies - the counter of pod failed is
+	// incremented and it is checked against the backoffLimit
+	// +optional
+	BackoffPolicy *BackoffPolicy
+  ...
+```
+
+Additionally, we validate the following constraints for each instance of `BackoffPolicyRule`:
+- exactly one of the following fields is used:`ContainerExitCodeIn`,
+ `ContainerExitCodeNotIn`, `PodStatusReasonIn`, `PodStatusReasonNotIn`.
+- the `ContainerName` is only combined with the fields `ContainerExitCodeIn` or
+ `ContainerExitCodeNotIn`.
+
+Here is an example Job configuration which uses this API:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: main-job-container
+        image: job-image
+        command: ["./program"]
+      - name: monitoring-job-container
+        image: job-monitoring
+        command: ["./monitoring"]
+  backoffLimit: 3
+  backoffPolicy:
+    rules:
+    - containerExitCodeIn: 1-12,100-127
+      containerName: main-job-container
+      action: Terminate
+    - podStatusReasonNotIn:
+      - ''
+      action: Ignore
+```
+
+<!--
+This is important to clarify: What does this mean?
+
+Would the rest of the pods be terminated? Would they allow to run? The simplest approach is to terminate all the pods, just like we do when the backoffLimit is reached. And we can introduce other options (and some specific to indexed jobs) in a follow up KEP next release.
+ -->
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
```

Comment:
You can say that the current job-controller has adequate testing for backoff. The SIG approver will say if that makes sense or not.